### PR TITLE
Introduce a `ReceiveMessageAdvice`

### PR DESCRIPTION
**Pull request by artembilan**
* Deprecate an `AbstractMessageSourceAdvice` in favor of
`default` method in the `MessageSourceMutator`
* Move a `applyReceiveOnlyAdviceChain()` logic into the `AbstractPollingEndpoint`:
now both `PollingConsumer` and `SourcePollingChannelAdapter` can use
`ReceiveMessageAdvice`
* Introduce a `SimpleActiveIdleReceiveMessageAdvice` based already
on the `ReceiveMessageAdvice` and deprecate a `SimpleActiveIdleMessageSourceAdvice`
which is fully replaceable with newly introduced `SimpleActiveIdleReceiveMessageAdvice`
* Add `@SuppressWarnings("deprecation")` for those out-of-the-box `ReceiveMessageAdvice`
implementation which still use an `AbstractMessageSourceAdvice` for
backward compatibility
* Document a new feature and give the `MessageSourceMutator` a new meaning

### CODE REVIEW

File: src/reference/asciidoc/polling-consumer.adoc [hunk 1]
```diff
@@ -145,26 +145,29 @@ These "`around advice`" methods do not have access to any context for the poll -
 This is fine for requirements such as making a task transactional or skipping a poll due to some external condition, as discussed earlier.
 What if we wish to take some action depending on the result of the `receive` part of the poll or if we want to adjust the poller depending on conditions? For those instances, Spring Integration offers "`Smart`" Polling.
 
+[[smart-polling]]
 ===== "`Smart`" Polling
 
-Version 4.2 introduced the `AbstractMessageSourceAdvice`.
-Any `Advice` objects in the `advice-chain` that subclass this class are applied only to the receive operation.
+Version 5.3 introduced the `ReceiveMessageAdvice` interface.
+(An `AbstractMessageSourceAdvice` has been deprecated in favor of `default` methods in the `MessageSourceMutator`.)
```
Comment (garyrussell) [inline src/reference/asciidoc/polling-consumer.adoc:152]: (The

File: spring-integration-core/src/main/java/org/springframework/integration/aop/CompoundTriggerAdvice.java [hunk 2]
```diff
@@ -47,8 +51,21 @@ public CompoundTriggerAdvice(CompoundTrigger compoundTrigger, Trigger overrideTr
...
+	 */
 	@Override
+	@Deprecated
 	public Message<?> afterReceive(Message<?> result, MessageSource<?> source) {
+		return afterReceive(result, (Object) source);
+	}
+
+	@Override
+	@Nullable
+	public Message<?> afterReceive(@Nullable Message<?> result, Object source) {
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/aop/CompoundTriggerAdvice.java:68]: Wouldn't something like `Advisable<T>` be better than `Object`?

**Review by garyrussell — CHANGES_REQUESTED** [review]

File: spring-integration-core/src/main/java/org/springframework/integration/aop/CompoundTriggerAdvice.java (hunk 2 above)
Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/aop/CompoundTriggerAdvice.java:68]: Well, in most cases it is really about both `MessageSource` and `PollableChannel`, so even with generic argument it still going to be an `Object` 😄 

I thought about a `Supplier` for both those interfaces, but this is not what we are going to use with the `receive(timeout)`.
Therefore I'm OK to stick with an `Object` for a `source` argument.

File: src/reference/asciidoc/polling-consumer.adoc (hunk 1 above)
Comment (artembilan) [inline src/reference/asciidoc/polling-consumer.adoc:152]:
```suggestion
(The `AbstractMessageSourceAdvice` has been deprecated in favor of `default` methods in the `MessageSourceMutator`.)
```

**Pull request merged**